File: c/allergies/src/allergies.c
#include "allergies.h"
#include <stdlib.h>

bool is_allergic_to(allergen_t allergen, int score) {
    return score & (1 << allergen);
}

void get_allergens(int score, allergen_list_t *list) {
    list->count = 0;
    list->allergens = malloc(ALLERGEN_COUNT * sizeof(allergen_t));
    for (allergen_t allergen = 0; allergen < ALLERGEN_COUNT; allergen++) {
        if (is_allergic_to(allergen, score)) {
	    list->allergens[list->count++] = allergen;
	}
    }
}

File: java/allergies/src/main/java/Allergies.java
import java.util.List;
import java.util.stream.Stream;
import java.util.stream.Collectors;

class Allergies {
    int score;

    Allergies(int score) {
        this.score = score;
    }

    boolean isAllergicTo(Allergen allergen) {
        return (score & allergen.getScore()) > 0;
    }

    List<Allergen> getList() {
        return Stream.of(Allergen.values())
            .filter(this::isAllergicTo)
            .collect(Collectors.toList());
    }
}
